File: App/app/src/main/java/com/cinema/ticket_booking/ui/booking/SelectSeatFragment.java
package com.cinema.ticket_booking.ui.booking;

import android.os.Bundle;
import android.view.*;
import android.widget.*;
import androidx.annotation.*;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.*;
import com.cinema.ticket_booking.R;
import com.cinema.ticket_booking.data.model.response.*;
import com.cinema.ticket_booking.util.SnackbarHelper;
import com.cinema.ticket_booking.databinding.FragmentSelectSeatBinding;
import java.util.*;
import java.text.SimpleDateFormat;
import java.util.Locale;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class SelectSeatFragment extends Fragment {

    private FragmentSelectSeatBinding binding;
    private SelectSeatViewModel viewModel;
    private String showtimeId;
    private SeatMapResponse currentSeatMap;
    private long expireTime = 0;

    // Polling for real-time seat status updates
    private final android.os.Handler pollHandler = new android.os.Handler(android.os.Looper.getMainLooper());
    private Runnable pollRunnable;
    private static final long POLL_INTERVAL_MS = 12_000; // 12 seconds

    // Map of showtimeSeatId -> seat button view (for efficient targeted updates)
    private final Map<String, TextView> seatButtonMap = new HashMap<>();

    @Override
    public View onCreateView(@NonNull LayoutInflater i, ViewGroup c, Bundle s) {
        binding = FragmentSelectSeatBinding.inflate(i, c, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewModel = new ViewModelProvider(this).get(SelectSeatViewModel.class);

        if (getArguments() != null) {
            showtimeId = getArguments().getString("showtimeId");
            viewModel.loadSeatMap(showtimeId);
        }

        String rawTime = SelectShowtimeViewModel.pendingShowtimeTime;
        String displayTime = rawTime;
        if (rawTime != null) {
            try {
                SimpleDateFormat sdfIn;
                if (rawTime.contains("T")) {
                    sdfIn = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.getDefault());
                } else {
                    sdfIn = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault());
                }
                Date d = sdfIn.parse(rawTime);
                SimpleDateFormat sdfOut = new SimpleDateFormat("HH:mm - dd/MM/yyyy", Locale.getDefault());
                displayTime = sdfOut.format(d);
            } catch (Exception e) {
                // Ignore
            }
        }

        binding.tvShowtimeInfo.setText(
                SelectShowtimeViewModel.pendingMovieTitle + " • " + displayTime);

        binding.btnBack.setOnClickListener(v -> {
            Navigation.findNavController(view).popBackStack();
        });

        binding.btnConfirm.setOnClickListener(v -> {
            if (viewModel.getSelectedSeatIds().isEmpty()) {
                SnackbarHelper.showError(binding.getRoot(), "Vui lòng chọn ít nhất 1 ghế");
                return;
            }
            SelectSeatViewModel.pendingSeatIds = new ArrayList<>(viewModel.getSelectedSeatIds());
            SelectSeatViewModel.pendingTotalAmount = viewModel.calculateTotal(currentSeatMap);

            Bundle bundle = new Bundle();
            bundle.putLong("expireTime", expireTime);
            Navigation.findNavController(view).navigate(R.id.action_selectSeat_to_selectCombo, bundle);
        });

        viewModel.getSeatMap().observe(getViewLifecycleOwner(), resource -> {
            switch (resource.status) {
                case LOADING -> binding.progressBar.setVisibility(View.VISIBLE);
                case SUCCESS -> {
                    binding.progressBar.setVisibility(View.GONE);
                    if (resource.data != null) {
                        currentSeatMap = resource.data;
                        expireTime = System.currentTimeMillis() + (long) resource.data.getSeatHoldMins() * 60 * 1000;
                        seatButtonMap.clear();
                        renderSeatMap(resource.data);
                        startPolling();
                    }
                }
                case ERROR -> {
                    binding.progressBar.setVisibility(View.GONE);
                    SnackbarHelper.showError(binding.getRoot(), resource.message);
                }
            }
        });

        // Silent refresh observer — only updates changed seat colors
        viewModel.getSeatRefresh().observe(getViewLifecycleOwner(), freshMap -> {
            if (freshMap == null || freshMap.getSeats() == null) return;
            currentSeatMap = freshMap;
            boolean anyDeselected = false;
            for (SeatMapResponse.SeatItem seat : freshMap.getSeats()) {
                TextView btn = seatButtonMap.get(seat.getShowtimeSeatId());
                if (btn == null) continue;
                boolean isSelectedByMe = viewModel.getSelectedSeatIds().contains(seat.getShowtimeSeatId());
                // If seat became taken but user had selected it, deselect
                if (isSelectedByMe && ("BOOKED".equals(seat.getStatus()) || "LOCKED".equals(seat.getStatus()))) {
                    viewModel.getSelectedSeatIds().remove(seat.getShowtimeSeatId());
                    isSelectedByMe = false;
                    anyDeselected = true;
                }
                updateSeatButtonStyle(btn, seat, isSelectedByMe);
            }
            if (anyDeselected) {
                SnackbarHelper.showError(binding.getRoot(), "Một số ghế bạn chọn vừa bị người khác đặt mất!");
                updateSummary();
            }
        });
    }

    private void startPolling() {
        if (pollRunnable != null) pollHandler.removeCallbacks(pollRunnable);
        pollRunnable = new Runnable() {
            @Override
            public void run() {
                if (binding != null) {
                    viewModel.refreshSeatStatuses();
                    pollHandler.postDelayed(this, POLL_INTERVAL_MS);
                }
            }
        };
        pollHandler.postDelayed(pollRunnable, POLL_INTERVAL_MS);
    }

    private void renderSeatMap(SeatMapResponse seatMap) {
        binding.seatContainer.removeAllViews();
        if (seatMap.getSeats() == null)
            return;

        int totalRows = seatMap.getMaxGridRow() + 1;
        int totalCols = seatMap.getMaxGridCol() + 1;

        // Build lookup map: "gridRow_gridCol" -> SeatItem
        Map<String, SeatMapResponse.SeatItem> seatGrid = new HashMap<>();
        for (SeatMapResponse.SeatItem seat : seatMap.getSeats()) {
            seatGrid.put(seat.getGridRow() + "_" + seat.getGridCol(), seat);
        }

        int seatSize = (int) (getResources().getDisplayMetrics().density * 34);
        int margin = (int) (getResources().getDisplayMetrics().density * 3);

        for (int r = 0; r < totalRows; r++) {
            // Determine row label: scan seats in this row to find the label char
            String label = String.valueOf((char) ('A' + r));
            for (int c = 0; c < totalCols; c++) {
                SeatMapResponse.SeatItem s = seatGrid.get(r + "_" + c);
                if (s != null && s.getSeatLabel() != null && !s.getSeatLabel().isEmpty()) {
                    label = String.valueOf(s.getSeatLabel().charAt(0));
                    break;
                }
            }
            // Always render the row (including empty rows) so the seat map is consistent

            LinearLayout row = new LinearLayout(requireContext());
            row.setOrientation(LinearLayout.HORIZONTAL);
            row.setGravity(android.view.Gravity.CENTER);
            row.setPadding(0, margin, 0, margin);

            // Left label
            TextView rowLabelLeft = new TextView(requireContext());
            rowLabelLeft.setText(label);
            rowLabelLeft.setTextColor(getResources().getColor(R.color.on_surface_variant, null));
            rowLabelLeft.setTypeface(null, android.graphics.Typeface.BOLD);
            rowLabelLeft.setMinWidth((int) (getResources().getDisplayMetrics().density * 40));
            rowLabelLeft.setGravity(android.view.Gravity.CENTER);
            row.addView(rowLabelLeft);

            for (int c = 0; c < totalCols; c++) {
                SeatMapResponse.SeatItem seat = seatGrid.get(r + "_" + c);

                if (seat != null) {
                    TextView seatBtn = new TextView(requireContext());
                    String seatText = seat.getSeatLabel() != null && seat.getSeatLabel().length() > 1
                            ? seat.getSeatLabel().substring(1)
                            : String.valueOf(seat.getColNumber());

                    seatBtn.setText(seatText);
                    seatBtn.setGravity(android.view.Gravity.CENTER);
                    seatBtn.setTextSize(10f);

                    LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams(seatSize, seatSize);
                    lp.setMargins(margin, 0, margin, 0);
                    seatBtn.setLayoutParams(lp);

                    // Track this button for real-time updates
                    seatButtonMap.put(seat.getShowtimeSeatId(), seatBtn);

                    boolean isInitiallySelected = viewModel.getSelectedSeatIds().contains(seat.getShowtimeSeatId());
                    updateSeatButtonStyle(seatBtn, seat, isInitiallySelected);

                    seatBtn.setOnClickListener(v -> {
                        boolean toggled = viewModel.toggleSeat(seat);
                        if (toggled) {
                            boolean isSelected = viewModel.getSelectedSeatIds().contains(seat.getShowtimeSeatId());
                            updateSeatButtonStyle(seatBtn, seat, isSelected);
                            updateSummary();
                        }
                    });
                    row.addView(seatBtn);
                } else {
                    View spacer = new View(requireContext());
                    LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams(seatSize, seatSize);
                    lp.setMargins(margin, 0, margin, 0);
                    spacer.setLayoutParams(lp);
                    row.addView(spacer);
                }
            }
            // Right label
            TextView rowLabelRight = new TextView(requireContext());
            rowLabelRight.setText(label);
            rowLabelRight.setTextColor(getResources().getColor(R.color.on_surface_variant, null));
            rowLabelRight.setMinWidth((int) (getResources().getDisplayMetrics().density * 40));
            rowLabelRight.setGravity(android.view.Gravity.CENTER);
            row.addView(rowLabelRight);

            binding.seatContainer.addView(row);
        }
        updateSummary();

        binding.zoomLayout.post(() -> {
            binding.zoomLayout.zoomTo(1.0f, false);
        });
    }

    private void updateSeatButtonStyle(TextView tv, SeatMapResponse.SeatItem seat, boolean selected) {
        int colorRes;
        if (selected)
            colorRes = R.color.seat_selected;
        else if ("BOOKED".equals(seat.getStatus()) || "LOCKED".equals(seat.getStatus()))
            colorRes = R.color.seat_booked;
        else if ("VIP".equals(seat.getSeatType()))
            colorRes = R.color.seat_vip;
        else if ("COUPLE".equals(seat.getSeatType()))
            colorRes = R.color.seat_couple;
        else
            colorRes = R.color.seat_available;

        int color = getResources().getColor(colorRes, null);

        android.graphics.drawable.GradientDrawable gd = new android.graphics.drawable.GradientDrawable();
        gd.setColor(color);
        float radius = getResources().getDisplayMetrics().density * 6;
        gd.setCornerRadius(radius);

        if (!selected && !"BOOKED".equals(seat.getStatus())) {
            gd.setStroke(2, getResources().getColor(R.color.outline_variant, null));
        }

        tv.setBackground(gd);
        tv.setTextColor(getResources().getColor(android.R.color.white, null));
    }

    private void updateSummary() {
        int count = viewModel.getSelectedSeatIds().size();
        double total = viewModel.calculateTotal(currentSeatMap);
        binding.tvSelectedCount.setText(count + " ghế đã chọn");
        binding.tvTotalPrice.setText(String.format("%,.0fđ", total));
        binding.btnConfirm.setEnabled(count > 0);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (pollRunnable != null) {
            pollHandler.removeCallbacks(pollRunnable);
        }
        binding = null;
    }
}

File: App/app/src/main/java/com/cinema/ticket_booking/ui/booking/SelectSeatFragment.kt
package com.cinema.ticket_booking.ui.booking

import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation
import com.cinema.ticket_booking.R
import com.cinema.ticket_booking.data.model.response.SeatMapResponse
import com.cinema.ticket_booking.databinding.FragmentSelectSeatBinding
import com.cinema.ticket_booking.util.Resource
import com.cinema.ticket_booking.util.SnackbarHelper
import dagger.hilt.android.AndroidEntryPoint
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

@AndroidEntryPoint
class SelectSeatFragment : Fragment() {

    private var _binding: FragmentSelectSeatBinding? = null
    private val binding get() = _binding!!
    private lateinit var viewModel: SelectSeatViewModel
    private var showtimeId: String? = null
    private var currentSeatMap: SeatMapResponse? = null
    private var expireTime: Long = 0

    // Polling for real-time seat status updates
    private val pollHandler = Handler(Looper.getMainLooper())
    private var pollRunnable: Runnable? = null

    // Map of showtimeSeatId -> seat button view (for efficient targeted updates)
    private val seatButtonMap = HashMap<String, TextView>()

    companion object {
        private const val POLL_INTERVAL_MS = 12_000L // 12 seconds
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentSelectSeatBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel = ViewModelProvider(this)[SelectSeatViewModel::class.java]

        arguments?.let {
            showtimeId = it.getString("showtimeId")
            showtimeId?.let { id -> viewModel.loadSeatMap(id) }
        }

        val rawTime = SelectShowtimeViewModel.pendingShowtimeTime
        var displayTime = rawTime ?: ""
        if (rawTime != null) {
            try {
                val sdfIn = if (rawTime.contains("T")) {
                    SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.getDefault())
                } else {
                    SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
                }
                val d = sdfIn.parse(rawTime)
                if (d != null) {
                    val sdfOut = SimpleDateFormat("HH:mm - dd/MM/yyyy", Locale.getDefault())
                    displayTime = sdfOut.format(d)
                }
            } catch (e: Exception) {
                // Ignore
            }
        }

        binding.tvShowtimeInfo.text = "${SelectShowtimeViewModel.pendingMovieTitle ?: ""} • $displayTime"

        binding.btnBack.setOnClickListener {
            Navigation.findNavController(view).popBackStack()
        }

        binding.btnConfirm.setOnClickListener {
            if (viewModel.selectedSeatIds.isEmpty()) {
                SnackbarHelper.showError(binding.root, "Vui lòng chọn ít nhất 1 ghế")
                return@setOnClickListener
            }
            SelectSeatViewModel.pendingSeatIds = ArrayList(viewModel.selectedSeatIds)
            SelectSeatViewModel.pendingTotalAmount = viewModel.calculateTotal(currentSeatMap)

            val bundle = Bundle().apply {
                putLong("expireTime", expireTime)
            }
            Navigation.findNavController(view).navigate(R.id.action_selectSeat_to_selectCombo, bundle)
        }

        viewModel.seatMap.observe(viewLifecycleOwner) { resource ->
            when (resource.status) {
                Resource.Status.LOADING -> binding.progressBar.visibility = View.VISIBLE
                Resource.Status.SUCCESS -> {
                    binding.progressBar.visibility = View.GONE
                    resource.data?.let { data ->
                        currentSeatMap = data
                        expireTime = System.currentTimeMillis() + data.seatHoldMins.toLong() * 60 * 1000
                        seatButtonMap.clear()
                        renderSeatMap(data)
                        startPolling()
                    }
                }
                Resource.Status.ERROR -> {
                    binding.progressBar.visibility = View.GONE
                    SnackbarHelper.showError(binding.root, resource.message ?: "Lỗi tải bản đồ ghế")
                }
            }
        }

        // Silent refresh observer — only updates changed seat colors
        viewModel.seatRefresh.observe(viewLifecycleOwner) { freshMap ->
            if (freshMap?.seats == null) return@observe
            currentSeatMap = freshMap
            var anyDeselected = false
            for (seat in freshMap.seats) {
                val seatId = seat.showtimeSeatId ?: continue
                val btn = seatButtonMap[seatId] ?: continue
                var isSelectedByMe = viewModel.selectedSeatIds.contains(seatId)
                // If seat became taken but user had selected it, deselect
                if (isSelectedByMe && ("BOOKED" == seat.status || "LOCKED" == seat.status)) {
                    viewModel.selectedSeatIds.remove(seatId)
                    isSelectedByMe = false
                    anyDeselected = true
                }
                updateSeatButtonStyle(btn, seat, isSelectedByMe)
            }
            if (anyDeselected) {
                SnackbarHelper.showError(binding.root, "Một số ghế bạn chọn vừa bị người khác đặt mất!")
                updateSummary()
            }
        }
    }

    private fun startPolling() {
        pollRunnable?.let { pollHandler.removeCallbacks(it) }
        pollRunnable = object : Runnable {
            override fun run() {
                if (_binding != null) {
                    viewModel.refreshSeatStatuses()
                    pollHandler.postDelayed(this, POLL_INTERVAL_MS)
                }
            }
        }
        pollRunnable?.let { pollHandler.postDelayed(it, POLL_INTERVAL_MS) }
    }

    private fun renderSeatMap(seatMap: SeatMapResponse) {
        binding.seatContainer.removeAllViews()
        if (seatMap.seats == null) return

        val totalRows = seatMap.maxGridRow + 1
        val totalCols = seatMap.maxGridCol + 1

        // Build lookup map: "gridRow_gridCol" -> SeatItem
        val seatGrid = HashMap<String, SeatMapResponse.SeatItem>()
        for (seat in seatMap.seats) {
            seatGrid["${seat.gridRow}_${seat.gridCol}"] = seat
        }

        val seatSize = (resources.displayMetrics.density * 34).toInt()
        val margin = (resources.displayMetrics.density * 3).toInt()

        for (r in 0 until totalRows) {
            // Determine row label: scan seats in this row to find the label char
            var label = ('A' + r).toString()
            for (c in 0 until totalCols) {
                val s = seatGrid["${r}_${c}"]
                if (s != null && !s.seatLabel.isNullOrEmpty()) {
                    label = s.seatLabel.substring(0, 1)
                    break
                }
            }
            // Always render the row (including empty rows) so the seat map is consistent

            val row = LinearLayout(requireContext()).apply {
                orientation = LinearLayout.HORIZONTAL
                gravity = Gravity.CENTER
                setPadding(0, margin, 0, margin)
            }

            // Left label
            val rowLabelLeft = TextView(requireContext()).apply {
                text = label
                setTextColor(resources.getColor(R.color.on_surface_variant, null))
                setTypeface(null, android.graphics.Typeface.BOLD)
                minWidth = (resources.displayMetrics.density * 40).toInt()
                gravity = Gravity.CENTER
            }
            row.addView(rowLabelLeft)

            for (c in 0 until totalCols) {
                val seat = seatGrid["${r}_${c}"]

                if (seat != null) {
                    val seatBtn = TextView(requireContext())
                    val seatText = if (seat.seatLabel != null && seat.seatLabel.length > 1) {
                        seat.seatLabel.substring(1)
                    } else {
                        seat.colNumber.toString()
                    }

                    seatBtn.text = seatText
                    seatBtn.gravity = Gravity.CENTER
                    seatBtn.textSize = 10f

                    val lp = LinearLayout.LayoutParams(seatSize, seatSize).apply {
                        setMargins(margin, 0, margin, 0)
                    }
                    seatBtn.layoutParams = lp

                    val seatId = seat.showtimeSeatId ?: ""
                    // Track this button for real-time updates
                    seatButtonMap[seatId] = seatBtn

                    val isInitiallySelected = viewModel.selectedSeatIds.contains(seatId)
                    updateSeatButtonStyle(seatBtn, seat, isInitiallySelected)

                    seatBtn.setOnClickListener {
                        val toggled = viewModel.toggleSeat(seat)
                        if (toggled) {
                            val isSelected = viewModel.selectedSeatIds.contains(seatId)
                            updateSeatButtonStyle(seatBtn, seat, isSelected)
                            updateSummary()
                        }
                    }
                    row.addView(seatBtn)
                } else {
                    val spacer = View(requireContext())
                    val lp = LinearLayout.LayoutParams(seatSize, seatSize).apply {
                        setMargins(margin, 0, margin, 0)
                    }
                    spacer.setLayoutParams(lp)
                    row.addView(spacer)
                }
            }
            // Right label
            val rowLabelRight = TextView(requireContext()).apply {
                text = label
                setTextColor(resources.getColor(R.color.on_surface_variant, null))
                minWidth = (resources.displayMetrics.density * 40).toInt()
                gravity = Gravity.CENTER
            }
            row.addView(rowLabelRight)

            binding.seatContainer.addView(row)
        }
        updateSummary()

        binding.zoomLayout.post {
            binding.zoomLayout.zoomTo(1.0f, false)
        }
    }

    private fun updateSeatButtonStyle(tv: TextView, seat: SeatMapResponse.SeatItem, selected: Boolean) {
        val colorRes = if (selected) {
            R.color.seat_selected
        } else if ("BOOKED" == seat.status || "LOCKED" == seat.status) {
            R.color.seat_booked
        } else if ("VIP" == seat.seatType) {
            R.color.seat_vip
        } else if ("COUPLE" == seat.seatType) {
            R.color.seat_couple
        } else {
            R.color.seat_available
        }

        val color = resources.getColor(colorRes, null)

        val gd = GradientDrawable().apply {
            setColor(color)
            val radius = resources.displayMetrics.density * 6
            cornerRadius = radius
        }

        if (!selected && "BOOKED" != seat.status) {
            gd.setStroke(2, resources.getColor(R.color.outline_variant, null))
        }

        tv.background = gd
        tv.setTextColor(resources.getColor(android.R.color.white, null))
    }

    private fun updateSummary() {
        val count = viewModel.selectedSeatIds.size
        val total = viewModel.calculateTotal(currentSeatMap)
        binding.tvSelectedCount.text = "$count ghế đã chọn"
        binding.tvTotalPrice.text = String.format("%,.0fđ", total)
        binding.btnConfirm.isEnabled = count > 0
    }

    override fun onDestroyView() {
        super.onDestroyView()
        pollRunnable?.let { pollHandler.removeCallbacks(it) }
        _binding = null
    }
}
